File: src/interpreter.js
/*
* Author: Jonas Finnemann Jensen
* Github: https://github.com/jonasfj
*/
var PrattParser = require('./prattparser');
var {isString, isNumber, isInteger, isBool,
  isArray, isObject, isFunction, isTruthy} = require('./type-utils');
var {InterpreterError} = require('./error');

let expectationError = (operator, expectation) => new InterpreterError(`${operator} expects ${expectation}`);

let isEqual = (a, b) =>  {
  if (isArray(a) && isArray(b) && a.length === b.length) {
    for (let i = 0; i < a.length; i++) {
      if (!isEqual(a[i], b[i])) { return false; }
    }
    return true;
  }
  if (isFunction(a)) {
    return a === b;
  }
  if (isObject(a) && isObject(b)) {
    let keys = Object.keys(a).sort();
    if (!isEqual(keys, Object.keys(b).sort())) { return false; }
    for (let k of keys) {
      if (!isEqual(a[k], b[k])) { return false; }
    }
    return true;
  }
  return a === b;
};

let parseList = (ctx, separator, terminator) => {
  let list = [];
  if (!ctx.attempt(terminator)) {
    do {
      list.push(ctx.parse());
    } while (ctx.attempt(separator));
    ctx.require(terminator);
  }
  return list;
};

let parseObject = (ctx) => {
  let obj = {};
  if (!ctx.attempt('}')) {
    do {
      let k = ctx.require('identifier', 'string');
      if (k.kind === 'string') {
        k.value = parseString(k.value);
      }
      ctx.require(':');
      let v = ctx.parse();
      obj[k.value] = v;
    } while (ctx.attempt(','));
    ctx.require('}');
  }
  return obj;
};

let parseInterval = (left, token, ctx) => {
  let a = null, b = null, isInterval = false;
  if (ctx.attempt(':')) {
    a = 0;
    isInterval = true;
  } else {
    a = ctx.parse();
    if (ctx.attempt(':')) {
      isInterval = true;
    }
  }

  if (isInterval && !ctx.attempt(']')) {
    b = ctx.parse();
    ctx.require(']');
  }

  if (!isInterval) {
    ctx.require(']');
  }

  return accessProperty(left, a, b, isInterval);
};

let accessProperty = (left, a, b, isInterval) => {
  if (isArray(left) || isString(left)) {
    if (isInterval) {
      b = b === null ? left.length : b;
      if (!isInteger(a) || !isInteger(b)) {
        throw new InterpreterError('cannot perform interval access with non-integers');
      }
      return left.slice(a, b);
    }
    if (!isInteger(a)) {
      throw new InterpreterError('should access arrays using integers only');
    }

    // for -ve index access
    a = a < 0 ? (left.length + a) % left.length : a;
    if (a >= left.length) {
      throw new InterpreterError('index out of bounds');
    }
    return left[a];
  }

  // if we reach here it means we are accessing property value from object
  if (!isObject(left)) {
    throw new InterpreterError('cannot access properties from non-objects');
  }

  if (!isString(a)) {
    throw new InterpreterError('object keys must be strings');
  }

  if (left.hasOwnProperty(a)) {
    return left[a];
  } else {
    return null;
  }
};

let parseString = (str) => {
  return str.slice(1, -1);
};

let testComparisonOperands = (operator, left, right) => {

  if (operator === '==' || operator === '!=') {
    return null;
  }

  let test = ['>=', '<=', '<', '>'].some(v => v === operator)
              && (isNumber(left) && isNumber(right) || isString(left) && isString(right));

  if (!test) {
    throw expectationError(`infix: ${operator}`, `numbers/strings ${operator} numbers/strings`);
  }
  return;
};

let testMathOperands = (operator, left, right) => {
  if (operator === '+' && !(isNumber(left) && isNumber(right) || isString(left) && isString(right))) {
    throw expectationError('infix: +', 'number/string + number/string');
  }
  if (['-', '*', '/', '**'].some(v => v === operator) && !(isNumber(left) && isNumber(right))) {
    throw expectationError(`infix: ${operator}`, `number ${operator} number`);
  }
  return;
};

let testLogicalOperand = (operator, operand) => {
  if (!isBool(operand)) {
    throw expectationError(`infix: ${operator}`, `boolean ${operator} boolean`);
  }
};

let prefixRules = {};
let infixRules = {};

// defining prefix rules
prefixRules['number'] = (token, ctx) => {
  let v = Number(token.value);
  if (isNaN(v)) {
    throw new Error(`${token.value} should be a number`);
  }
  return v;
};

prefixRules['!'] = (token, ctx) => {
  let operand = ctx.parse('unary');
  return !isTruthy(operand);
};

prefixRules['-'] = (token, ctx) => {
  let v = ctx.parse('unary');

  if (!isNumber(v)) {
    throw expectationError('unary -', 'number');
  }

  return -v;
};

prefixRules['+'] = (token, ctx) => {
  let v = ctx.parse('unary');

  if (!isNumber(v)) {
    throw expectationError('unary +', 'number');
  }

  return +v;
};

prefixRules['identifier'] = (token, ctx) => {
  if (ctx.context.hasOwnProperty(token.value)) {
    return ctx.context[token.value];
  }
  throw new InterpreterError(`unknown context value ${token.value}`);
};

prefixRules['null'] = (token, ctx) => {
  return null;
};

prefixRules['['] = (token, ctx) => parseList(ctx, ',', ']');

prefixRules['('] = (token, ctx) => {
  let v = ctx.parse();
  ctx.require(')');
  return v;
};

prefixRules['{'] = (token, ctx) => parseObject(ctx);

prefixRules['string'] = (token, ctx) => parseString(token.value);

prefixRules['true'] = (token, ctx) => {
  if (token.value === 'true') {
    return true;
  }
  throw new Error('Only \'true/false\' is considered as bool');
};

prefixRules['false'] = (token, ctx) => {
  if (token.value === 'false') {
    return false;
  }
  throw new Error('Only \'true/false\' is considered as bool');
};

// infix rule definition starts here
infixRules['+'] = infixRules['-'] = infixRules['*'] = infixRules['/']
  = (left, token, ctx) => {
    let operator = token.kind;
    let right = ctx.parse(operator);
    testMathOperands(operator, left, right);
    switch (operator) {
      case '+':  return left + right;
      case '-':  return left - right;
      case '*':  return left * right;
      case '/':  return left / right;
      default: throw new Error(`unknown infix operator: '${operator}'`);
    }
  };

infixRules['**'] = (left, token, ctx) => {
  let operator = token.kind;
  let right = ctx.parse('**-right-associative');
  testMathOperands(operator, left, right);
  if (typeof left !== typeof right) {
    throw new InterpreterError(`TypeError: ${typeof left} ${operator} ${typeof right}`);
  }
  return Math.pow(left, right);
};

infixRules['['] = (left, token, ctx) => parseInterval(left, token, ctx);

infixRules['.'] = (left, token, ctx) => {
  if (isObject(left)) {
    let key = ctx.require('identifier').value;
    if (left.hasOwnProperty(key)) {
      return left[key];
    }
    throw new InterpreterError('object has no property ' + key);
  }
  throw expectationError('infix: .', 'objects');
};

infixRules['('] =  (left, token, ctx) => {
  if (isFunction(left)) {
    return left.apply(null, parseList(ctx, ',', ')'));
  }
  throw expectationError('infix: f(args)', 'f to be function');
};

infixRules['=='] = infixRules['!='] = infixRules['<='] =
infixRules['>='] = infixRules['<'] =  infixRules['>']
  =  (left, token, ctx) => {
    let operator = token.kind;
    let right = ctx.parse(operator);
    testComparisonOperands(operator, left, right);
    switch (operator) {
      case '>=': return left >= right;
      case '<=': return left <= right;
      case '>':  return left > right;
      case '<':  return left < right;
      case '==': return isEqual(left, right);
      case '!=': return !isEqual(left, right);
      default:   throw new Error('no rule for comparison operator: ' + operator);
    }
  };

infixRules['||'] = infixRules['&&'] = (left, token, ctx) => {
  let operator = token.kind;
  let right = ctx.parse(operator);
  switch (operator) {
    case '||':  return isTruthy(left) || isTruthy(right);
    case '&&':  return isTruthy(left) && isTruthy(right);
    default:    throw new Error('no rule for boolean operator: ' + operator);
  }
};

infixRules['in'] = (left, token, ctx) => {
  let right = ctx.parse(token.kind);
  if (isObject(right)) {
    if (!isString(left)) {
      throw expectationError('Infix: in-object', 'string on left side');
    }
    right = Object.keys(right);
  } else if (isString(right)) {
    if (!isString(left)) {
      throw expectationError('Infix: in-string', 'string on left side');
    }
    // short-circuit to indexOf since this is a substring operation
    return right.indexOf(left) !== -1;
  } else if (!isArray(right)) {
    throw expectationError('Infix: in', 'Array, string, or object on right side');
  }

  return right.some(r => isEqual(left, r));
};

module.exports = new PrattParser({
  ignore: '\\s+', // ignore all whitespace including \n
  patterns: {
    number:     '[0-9]+(?:\\.[0-9]+)?',
    identifier: '[a-zA-Z_][a-zA-Z_0-9]*',
    string:     '\'[^\']*\'|"[^"]*"',
    // avoid matching these as prefixes of identifiers e.g., `insinutations`
    true: 'true(?![a-zA-Z_0-9])',
    false: 'false(?![a-zA-Z_0-9])',
    in: 'in(?![a-zA-Z_0-9])',
    null: 'null(?![a-zA-Z_0-9])',
  },
  tokens: [
    '**', ...'+-*/[].(){}:,'.split(''),
    '>=', '<=', '<', '>', '==', '!=', '!', '&&', '||',
    'true', 'false', 'in', 'null', 'number',
    'identifier', 'string',
  ],
  precedence: [
    ['||'],
    ['&&'],
    ['in'],
    ['==', '!='],
    ['>=', '<=', '<', '>'],
    ['+', '-'],
    ['*', '/'],
    ['**-right-associative'],
    ['**'],
    ['[', '.'],
    ['('],
    ['unary'],
  ],
  prefixRules,
  infixRules,
});

File: src/interpreter.rs
#![allow(unused_variables)]
#![allow(dead_code)]
use crate::errors::Error;
use crate::prattparser::{Context, PrattParser};
use crate::tokenizer::Token;
use serde_json::{map::Map, Number, Value};
use std::collections::HashMap;

pub(crate) struct Interpreter {
    parser: PrattParser<'static, Value, HashMap<String, Value>>,
}

impl Interpreter {
    pub(crate) fn new() -> Interpreter {
        Interpreter {
            // this would only fail if the parser itself were malformed
            parser: create_interpreter().expect("Interpreter initialization failed"),
        }
    }

    /// Parse a string with the given context.
    pub(crate) fn parse(
        &self,
        source: &str,
        context: HashMap<String, Value>,
    ) -> Result<Value, Error> {
        self.parser.parse(source, context, 0)
    }
}

fn parse_list(
    context: &mut Context<Value, HashMap<String, Value>>,
    separator: &str,
    terminator: &str,
) -> Result<Value, Error> {
    let mut list: Vec<Value> = vec![];

    if context.attempt(|t| t == terminator)? == None {
        loop {
            list.push(context.parse(None)?);
            if context.attempt(|t| t == separator)? == None {
                break;
            }
        }
        context.require(|t| t == terminator)?;
    }

    Ok(Value::Array(list))
}

fn parse_object(context: &mut Context<Value, HashMap<String, Value>>) -> Result<Value, Error> {
    let mut obj = Map::new();

    if context.attempt(|t| t == "}")? == None {
        loop {
            let mut k = context.require(|t| t == "identifier" || t == "string")?;

            if k.token_type == "string" {
                k.value = &k.value[1..k.value.len() - 1];
            }
            context.require(|t| t == ":")?;
            let v = context.parse(None)?;
            obj.insert(k.value.to_string(), v);

            if context.attempt(|t| t == ",")? == None {
                break;
            }
        }
        context.require(|t| t == "}")?;
    }

    Ok(Value::Object(obj))
}

fn parse_string(string: &str) -> Result<Value, Error> {
    Ok(Value::String(string[1..string.len() - 1].into()))
}

/// Serde's Numbers aren't really meant for calculations, and as_i64 will fail if
/// the Number was constructed with from_f64, even if it is an integer f64.
fn number_to_i64(number: serde_json::Number) -> Option<i64> {
    if let Some(i) = number.as_i64() {
        return Some(i);
    }

    if let Some(i) = number.as_f64() {
        if i.fract() == 0f64 {
            // TODO: more chances for overflow here..
            return Some(i as i64);
        }
    }

    None
}

fn create_interpreter() -> Result<PrattParser<'static, Value, HashMap<String, Value>>, Error> {
    let mut patterns = HashMap::new();
    patterns.insert("number", "[0-9]+(?:\\.[0-9]+)?");
    patterns.insert("identifier", "[a-zA-Z_][a-zA-Z_0-9]*");
    patterns.insert("string", "\'[^\']*\'|\"[^\"]*\"");
    // avoid matching these as prefixes of identifiers e.g., `insinutations`
    patterns.insert("true", "true(?![a-zA-Z_0-9])");
    patterns.insert("false", "false(?![a-zA-Z_0-9])");
    patterns.insert("in", "in(?![a-zA-Z_0-9])");
    patterns.insert("null", "null(?![a-zA-Z_0-9])");

    let token_types = vec![
        "**",
        "+",
        "-",
        "*",
        "/",
        "[",
        "]",
        ".",
        "(",
        ")",
        "{",
        "}",
        ":",
        ",",
        ">=",
        "<=",
        "<",
        ">",
        "==",
        "!=",
        "!",
        "&&",
        "||",
        "true",
        "false",
        "in",
        "null",
        "number",
        "identifier",
        "string",
    ];

    let precedence = vec![
        vec!["||"],
        vec!["&&"],
        vec!["in"],
        vec!["==", "!="],
        vec![">=", "<=", "<", ">"],
        vec!["+", "-"],
        vec!["*", "/"],
        vec!["**-right-associative"],
        vec!["**"],
        vec!["[", "."],
        vec!["("],
        vec!["unary"],
    ];

    let mut prefix_rules: HashMap<
        &str,
        fn(&Token, &mut Context<Value, HashMap<String, Value>>) -> Result<Value, Error>,
    > = HashMap::new();

    prefix_rules.insert("number", |token, _context| {
        let n = serde_json::from_str(token.value)
            .map_err(|e| Error::SyntaxError(format!("Error parsing number: {}", e)))?;
        Ok(n)
    });

    prefix_rules.insert("!", |_token, context| {
        let operand = context.parse(Some("unary"))?;
        match operand {
            // TODO: use is_truthy
            Value::Null => Ok(Value::Bool(true)),
            Value::String(o) => Ok(Value::Bool(o.len() == 0)),
            Value::Number(o) => Ok(Value::Bool(o.as_f64() == Some(0f64))),
            Value::Array(o) => Ok(Value::Bool(o.len() == 0)),
            Value::Object(o) => Ok(Value::Bool(o.is_empty())),
            Value::Bool(o) => Ok(Value::Bool(!o)),
        }
    });

    prefix_rules.insert("-", |_token, context| {
        let v = context.parse(Some("unary"))?;
        if let Some(n) = v.as_f64() {
            return Ok(Value::Number(Number::from_f64(-n).unwrap()));
        } else {
            return Err(Error::InterpreterError(
                "This operator expects a number".to_string(),
            ));
        }
    });

    prefix_rules.insert("+", |_token, context| {
        let v = context.parse(Some("unary"))?;
        if let Some(n) = v.as_f64() {
            return Ok(Value::Number(Number::from_f64(n).unwrap()));
        } else {
            return Err(Error::InterpreterError(
                "This operator expects a number".to_string(),
            ));
        }
    });

    prefix_rules.insert("identifier", |token, context| {
        if let Some(v) = context.context.get(token.value) {
            return Ok(v.clone());
        }
        return Err(Error::InterpreterError(format!(
            "unknown context value {}",
            token.value
        )));
    });

    prefix_rules.insert("null", |_token, _context| Ok(Value::Null));

    prefix_rules.insert("[", |_token, context| parse_list(context, ",", "]"));

    prefix_rules.insert("(", |_token, context| {
        let expression = context.parse(None)?;

        context.require(|t| t == ")")?;

        Ok(expression)
    });

    prefix_rules.insert("{", |_token, context| parse_object(context));

    prefix_rules.insert("string", |token, _context| {
        Ok(Value::String(token.value[1..token.value.len() - 1].into()))
    });

    prefix_rules.insert("true", |_token, _context| Ok(Value::Bool(true)));

    prefix_rules.insert("false", |_token, _context| Ok(Value::Bool(false)));

    let mut infix_rules: HashMap<
        &str,
        fn(&Value, &Token, &mut Context<Value, HashMap<String, Value>>) -> Result<Value, Error>,
    > = HashMap::new();

    infix_rules.insert("+", |left, token, context| {
        let right = context.parse(Some("+"))?;

        match (&left, &right) {
            (&Value::Number(_), &Value::Number(_)) => {
                let sum = left.as_f64().unwrap() + right.as_f64().unwrap();
                Ok(Value::Number(Number::from_f64(sum.into()).unwrap()))
            }
            (&Value::String(_), &Value::String(_)) => {
                let sum = [left.as_str().unwrap(), right.as_str().unwrap()].concat();
                Ok(Value::String(sum))
            }
            (_, _) => Err(Error::InterpreterError(
                "infix: +', 'number/string + number/string".to_string(),
            )),
        }
    });

    infix_rules.insert("*", |left, token, context| {
        let right = context.parse(Some("*"))?;

        match (&left, &right) {
            (&Value::Number(_), &Value::Number(_)) => {
                let product = left.as_f64().unwrap() * right.as_f64().unwrap();
                Ok(Value::Number(Number::from_f64(product.into()).unwrap()))
            }
            (_, _) => Err(Error::InterpreterError(
                "infix: *', 'number + number".to_string(),
            )),
        }
    });

    infix_rules.insert("**", |left, token, context| {
        let right = context.parse(Some("**-right-associative"))?;

        match (&left, &right) {
            (&Value::Number(_), &Value::Number(_)) => {
                let result = left.as_f64().unwrap().powf(right.as_f64().unwrap());
                Ok(Value::Number(Number::from_f64(result.into()).unwrap()))
            }
            (_, _) => Err(Error::InterpreterError(
                "infix: **', 'number + number".to_string(),
            )),
        }
    });

    // todo infix [
    #[allow(unused_assignments)] // temporary
    infix_rules.insert("[", |left, token, context| {
        let mut a: i64;
        let b: i64;
        let mut is_interval = false;

        // Successful cases this function handles:
        //  [1,2][1] - integer index of array [DONE]
        //  [1,2][1:2] - interval of array
        //  "abc"[1] - integer index of string
        //  "abc"[1:2] - interval of string
        //  {bar: 10}['bar'] - object property access

        if context.attempt(|t| t == ":")?.is_some() {
            a = 0;
            is_interval = true;
        } else {
            let parsed = context.parse(None)?;
            print!("parsed is {:?}\n", parsed);
            if let Value::Number(n) = parsed {
                if let Some(int_val) = number_to_i64(n) {
                    a = int_val;
                } else {
                    return Err(Error::InterpreterError(
                        "should only use integers to access arrays or strings".to_string(),
                    ));
                }
                if context.attempt(|t| t == ":")?.is_some() {
                    is_interval = true;
                }
            } else {
                return Err(Error::InterpreterError(
                    "left part of slice operator is not a number".to_string(),
                ));
            }
        }

        if is_interval && !context.attempt(|t| t == "]")?.is_some() {
            let parsed = context.parse(None)?;
            if let Value::Number(n) = parsed {
                if let Some(int_val) = number_to_i64(n) {
                    b = int_val;
                } else {
                    return Err(Error::InterpreterError(
                        "cannot perform interval access with non-integers".to_string(),
                    ));
                }
            } else {
                return Err(Error::InterpreterError(
                    "right part of slice operator is not a number".to_string(),
                ));
            }

            context.require(|t| t == "]")?;
        }

        if is_interval {
            unreachable!()
        } else {
            match left {
                Value::Array(ref arr) => {
                    if a < 0 {
                        a = a + (arr.len() as i64);
                    }

                    if a >= (arr.len() as i64) || a < 0 {
                        print!("index is {}\n", a);
                        return Err(Error::InterpreterError("index out of bounds".to_string()));
                    }

                    return Ok(arr[a as usize].clone());
                }

                _ => unreachable!(),
            }
        }
    });

    // todo infix .
    // todo infix (
    // todo infix ==
    // todo infix >= <= > < == !=
    // todo infix ||
    // todo infix &&
    // todo infix in

    PrattParser::new(
        "\\s+",
        patterns,
        token_types,
        precedence,
        prefix_rules,
        infix_rules,
    )
}

#[cfg(test)]
mod tests {
    use crate::errors::Error;
    use crate::interpreter::create_interpreter;
    use serde_json::{json, map::Map, Value};
    use std::collections::HashMap;

    #[test]
    fn parse_number_expression() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(
            interpreter.parse("23.67", HashMap::new(), 0).unwrap(),
            23.67
        );
    }

    #[test]
    fn parse_boolean_negation() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("!true", HashMap::new(), 0).unwrap(),
            false
        );
    }

    #[test]
    fn parse_minus_expression_negative_number() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("-7", HashMap::new(), 0).unwrap(), -7.0);
    }

    #[test]
    fn parse_minus_expression_double_negative() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("--7", HashMap::new(), 0).unwrap(), 7.0);
    }

    #[test]
    fn parse_minus_expression_plus() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("-+10", HashMap::new(), 0).unwrap(), -10.0);
    }

    #[test]
    fn parse_minus_expression_zero() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("-0", HashMap::new(), 0).unwrap(), 0.0);
    }

    #[test]
    fn parse_plus_expression_positive_number() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("+5", HashMap::new(), 0).unwrap(), 5.0);
    }

    #[test]
    fn parse_plus_expression_zero() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("+0", HashMap::new(), 0).unwrap(), 0.0);
    }

    #[test]
    fn parse_plus_expression_minus() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("+-10", HashMap::new(), 0).unwrap(), -10.0);
    }

    #[test]
    fn parse_boolean_true() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(interpreter.parse("true", HashMap::new(), 0).unwrap(), true);
    }

    #[test]
    fn parse_boolean_false() {
        let interpreter = create_interpreter().unwrap();

        assert_eq!(
            interpreter.parse("false", HashMap::new(), 0).unwrap(),
            false
        );
    }

    #[test]
    fn parse_identifier_positive() {
        let interpreter = create_interpreter().unwrap();
        let mut context = HashMap::new();
        context.insert("x".to_string(), Value::Number(10.into()));

        assert_eq!(interpreter.parse("x", context, 0).unwrap(), 10);
    }

    #[test]
    fn parse_identifier_negative_non_existing_variable() {
        let interpreter = create_interpreter().unwrap();
        let mut context = HashMap::new();
        context.insert("x".to_string(), Value::Number(10.into()));

        assert_eq!(
            interpreter.parse("y", context, 0).err(),
            Some(Error::InterpreterError(
                "unknown context value y".to_string()
            ))
        );
    }

    #[test]
    fn parse_array_positive() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1, 2]", HashMap::new(), 0).unwrap(),
            Value::Array(vec![1.into(), 2.into()]),
        );
    }

    #[test]
    fn parse_array_empty() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[]", HashMap::new(), 0).unwrap(),
            Value::Array(vec![]),
        );
    }

    #[test]
    fn parse_parens_negative() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("(true]", HashMap::new(), 0).err(),
            Some(Error::SyntaxError("Unexpected token error".to_string())),
        );
    }

    #[test]
    fn parse_parens() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("2*(3+4)", HashMap::new(), 0).unwrap(),
            json!(14.0),
        );
    }

    #[test]
    fn parse_curly_braces() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("{a: 10}", HashMap::new(), 0).unwrap(),
            json!({"a": 10}),
        );
    }

    #[test]
    fn parse_curly_braces_negative_semicolon() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("{b: 20; a: 10}", HashMap::new(), 0).err(),
            Some(Error::SyntaxError(
                "unexpected EOF for {b: 20; a: 10} at ; a: 10}".to_string()
            )),
        );
    }

    #[test]
    fn parse_curly_braces_two_keys() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter
                .parse("{b: 20, a: 10}", HashMap::new(), 0)
                .unwrap(),
            json!({"a": 10, "b": 20}),
        );
    }

    #[test]
    fn parse_curly_braces_empty() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("{}", HashMap::new(), 0).unwrap(),
            Value::Object(Map::new()),
        );
    }

    #[test]
    fn parse_curly_braces_string() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter
                .parse("{\"abc def\": 10}", HashMap::new(), 0)
                .unwrap(),
            json!({"abc def": 10}),
        );
    }

    #[test]
    fn parse_string() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("\"banana\"", HashMap::new(), 0).unwrap(),
            Value::String("banana".to_string()),
        );
    }

    #[test]
    fn parse_string_empty() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("\"\"", HashMap::new(), 0).unwrap(),
            Value::String("".to_string()),
        );
    }

    #[test]
    fn parse_string_addition() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter
                .parse("\"banana\" + \"chocolate\"", HashMap::new(), 0)
                .unwrap(),
            Value::String("bananachocolate".to_string()),
        );
    }

    #[test]
    fn parse_exponentiation() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("2**3", HashMap::new(), 0).unwrap(),
            json!(8.0),
        );
    }

    #[test]
    fn parse_exponentiation_zero_base() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("2**0", HashMap::new(), 0).unwrap(),
            json!(1.0),
        );
    }

    #[test]
    fn parse_array_of_integers() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2][1]", HashMap::new(), 0).unwrap(),
            json!(2),
        );
    }

    #[test]
    fn parse_array_negative_index() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2][-1]", HashMap::new(), 0).unwrap(),
            json!(2),
        );
    }

    #[test]
    fn parse_array_noninteger_index() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2][0.7]", HashMap::new(), 0).err(),
            Some(Error::InterpreterError(
                "should only use integers to access arrays or strings".to_string()
            ))
        );
    }

    #[test]
    fn parse_array_string_index() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2]['foo']", HashMap::new(), 0).err(),
            Some(Error::InterpreterError(
                // TODO: this is not the same message as JS
                "left part of slice operator is not a number".to_string()
            ))
        );
    }

    #[test]
    fn parse_array_positive_index_overflow() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2][9]", HashMap::new(), 0).err(),
            Some(Error::InterpreterError("index out of bounds".to_string()))
        );
    }

    #[test]
    fn parse_array_negative_index_overflow() {
        let interpreter = create_interpreter().unwrap();
        assert_eq!(
            interpreter.parse("[1,2][-9]", HashMap::new(), 0).err(),
            Some(Error::InterpreterError("index out of bounds".to_string()))
        );
    }
}
